heron/healthmgr/src/java/com/twitter/heron/healthmgr/HealthManagerMetrics.java

```diff
@@ -0,0 +1,181 @@
+   * @throws IOException
+   */
+  public HealthManagerMetrics(int metricsMgrPort) throws IOException {
+    jvmMetrics = new JVMMetrics();
+
+    executeSensorCount = new MultiCountMetric();
+    executeDetectorCount = new MultiCountMetric();
+    executeDiagnoserCount = new MultiCountMetric();
+    executeResolverCount = new MultiCountMetric();
+
+    looper = new NIOLooper();
+
+    SystemConfig systemConfig =
+        (SystemConfig) SingletonRegistry.INSTANCE.getSingleton(SystemConfig.HERON_SYSTEM_CONFIG);
+
+    HeronSocketOptions socketOptions =
+        new HeronSocketOptions(systemConfig.getInstanceNetworkWriteBatchSize(),
+            systemConfig.getInstanceNetworkWriteBatchTime(),
+            systemConfig.getInstanceNetworkReadBatchSize(),
+            systemConfig.getInstanceNetworkReadBatchTime(),
+            systemConfig.getInstanceNetworkOptionsSocketSendBufferSize(),
+            systemConfig.getInstanceNetworkOptionsSocketReceivedBufferSize(),
+            systemConfig.getInstanceNetworkOptionsMaximumPacketSize());
+    metricsMgrClient =
+        new MetricsMgrClient(looper, METRICS_MGR_HOST, metricsMgrPort, socketOptions);
+
+    outMetricsQueues = new Communicator<Metrics.MetricPublisherPublishMessage>(null, looper);
+    outMetricsQueues.init(systemConfig.getInstanceInternalMetricsWriteQueueCapacity(),
+        systemConfig.getInstanceTuningExpectedMetricsWriteQueueSize(),
+        systemConfig.getInstanceTuningCurrentSampleWeight());
+
+    int interval = (int) systemConfig.getHeronMetricsExportInterval().getSeconds();
+
+    looper.registerTimerEvent(Duration.ofSeconds(interval), new Runnable() {
+      @Override
+      public void run() {
+        jvmMetrics.getJVMSampleRunnable().run();
+
+        // push to container 0 metricsMgr
+        if (!metricsMgrClient.isConnected()) {
+          return;
+        }
+
+        LOG.info("Flushing all pending data in MetricsManagerClient");
+        // Collect all tuples in queue
+        int size = outMetricsQueues.size();
+        for (int i = 0; i < size; i++) {
+          Metrics.MetricPublisherPublishMessage m = outMetricsQueues.poll();
+          metricsMgrClient.sendMessage(m);
+        }
+      }
+    });
+  }
+
+  public Communicator<Metrics.MetricPublisherPublishMessage> getMetricsQueue() {
+    return outMetricsQueues;
+  }
+
+  public synchronized void executeSensorIncr(String sensor) {
+    executeSensorCount.scope(sensor);
+  }
+
+  public synchronized void executeDetectorIncr(String detector) {
+    executeDetectorCount.scope(detector);
+  }
+
+  public synchronized void executeDiagnoserIncr(String diagnoser) {
+    executeDiagnoserCount.scope(diagnoser);
+  }
+
+  public synchronized void executeResolver(String resolver) {
+    executeResolverCount.scope(resolver);
+  }
+
+  @Override
+  public void run() {
+    metricsMgrClient.start();
+    looper.loop();
+  }
+
+  @Override
+  public void close() throws Exception {
+    metricsMgrClient.stop();
```

Comment:
Might be safer to be in this order?

looper.exitLoop();
metricsMgrClient.stop();
outMetricsQueues.clear();

Reply:
done